M2-5203: Change password form and route

sultanofcardio opened the pull request:
### 📝 Description

🔗 [Jira Ticket M2-5203](https://mindlogger.atlassian.net/browse/M2-5203)

This PR adds a password recovery page that mirrors the same page in the respondent web app. Prior to this change, users of the admin app were not able to change their password without going to the respondent web app.

### 📸 Screenshots

Here's a screen recording of me going through the password recovery process

https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/ac76ddf7-d59c-4e45-88bd-b3c7abc5ffb7

> [!WARNING]
> The email sent from the API [currently links](https://github.com/ChildMindInstitute/mindlogger-backend-refactor/pull/1100) to a `/password-recovery` path in the admin app, but the correct path is `/auth/password-recovery`, so I had to edit the link before going through with it. We will have to update the API to fix this

### 🪤 Peer Testing

#### Happy path

- Start up the API and the admin panel
- Navigate to the admin app landing page
- Click the forgot password link
- Enter your email address and click *Send me a reset link*
- Copy the link in your email (Go to http://localhost:8025/ for local testing), and **edit** it to prepend `/auth` to the path
- Enter the new password
  **Expected outcome:** You should be redirected to the login screen and see a success banner displayed for 5 seconds

#### Invalid email/key

Enter each of the following URLs in your browser with the admin app running and confirm that you see an error message saying the link is invalid
- `https://localhost:3000/auth/password-recovery`
- `https://localhost:3000/auth/password-recovery?email=someone%40example.com&key=invalid=key`

Also repeat the test while the backend is not running

<img width="640" alt="image" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/544d70c6-5116-4aef-a7fb-c46ab15e05e2">

#### Error changing password

- Repeat the steps from the happy path until you arrive at the password change form
- Turn off the backend API
- Submit the form and confirm that you're seeing an appropriate error message

<img width="640" alt="image" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/c1bdc394-6957-4a2e-b261-a35cb2fd6010">

### ✏️ Notes

Be mindful that the email sent from the API [currently links](https://github.com/ChildMindInstitute/mindlogger-backend-refactor/pull/1100) to a `/password-recovery` path in the admin app, but the correct path is `/auth/password-recovery`. Be sure to edit the link in the email while testing to correct this

sultanofcardio committed:
M2-5203: Implement password recovery page

sultanofcardio commented:
P.S. I'm not sure yet why there are some tests failing but I'd like to get some eyes on this in the meantime

ChaconC commented on src/shared/components/Banners/Banners.const.ts:
```diff
@@ -1,6 +1,7 @@
 import { ComponentType } from 'react';
 
 import { BannerType } from 'redux/modules';
+import {PasswordResetSuccessfulBanner} from 'modules/Auth/features/Login/Banners';
```
Missing code formatting

sultanofcardio replied:
Ah, thanks for catching. In a few minutes (🤞) we should be able to catch these automatically through a GitHub action

ChaconC commented on src/modules/Auth/pages/RecoverPassword/RecoverPassword.tsx:
```diff
@@ -0,0 +1,59 @@
+              <StyledLinkBtn
+                style={{ marginBottom: 3, fontSize: 16 }}
+                variant="text"
+                onClick={() => navigate(page.login)}
```
I think this should link to `page.passwordReset`, right?
The text reads "Please click here to reset your password"

sultanofcardio replied:
That's exactly right. I must have messed up and pasted over this somewhere along the way. Thanks Carlos!

ChaconC commented on src/modules/Auth/pages/RecoverPassword/RecoverPassword.tsx:
```diff
@@ -0,0 +1,59 @@
+  } = useAsync(
+    async (data: { email: string | null; key: string | null }) =>
+      await recoveryLinkHealthCheckApi(data),
+  );
```
Granted it doesn't affect functionality, if you navigate to a used link this results in a couple of uncaught exceptions

sultanofcardio replied:
I ended up just calling the function inside a useEffect, and I handled the errors there

ChaconC reviewed (commented):
Couple of minor issues from me but overall works and looks great!

sultanofcardio committed:
M2-5203: Fix link to password reset page

sultanofcardio committed:
M2-4835: Linting

sultanofcardio committed:
M2-4835: Replace useAsync with plain useEffect

useAsync seems to be for callbacks, which doesn't really apply here

sultanofcardio committed:
M2-4835: Wrap spinner in Box so it becomes full page

alex-i-e commented on src/modules/Auth/features/RecoverPassword/RecoverForm/RecoverForm.schema.ts:
```diff
@@ -0,0 +1,27 @@
+      passwordConfirmation: yup
+        .string()
+        .required(passwordConfirmationRequired)
+        .oneOf([yup.ref('password')], passwordsMustMatch),
```
it's rare case, but anyway an user can be confused the cases when input message should be 'Please confirm your password' instead of 'Your passwords do not match':
![image](https://github.com/ChildMindInstitute/mindlogger-admin/assets/16320432/5f735519-477c-4abb-b332-5994891b8b5d)
![image](https://github.com/ChildMindInstitute/mindlogger-admin/assets/16320432/bb04cd97-c4ff-4f2b-ab00-db836f4ad720)

 I would expect to receive the message when Confirm Password is empty: 
![image](https://github.com/ChildMindInstitute/mindlogger-admin/assets/16320432/18cc8ba8-2e78-4e56-b7aa-05624ba7d04d)
![image](https://github.com/ChildMindInstitute/mindlogger-admin/assets/16320432/8c82e8b1-4174-4450-9ec7-79a2c49ed0c1)
```suggestion
        .oneOf([yup.ref('password'), ''], passwordsMustMatch),
```

sultanofcardio replied:
Good catch Aleksandr. Thanks for the suggestion

alex-i-e commented on src/modules/Auth/pages/RecoverPassword/RecoverPassword.tsx:
```diff
@@ -0,0 +1,62 @@
+  const key = searchParams.get('key');
+  const email = searchParams.get('email');
+
+  const [isLoading, setIsLoading] = useState<boolean>(true);
```
please consider to use `useAsync`, I believe this will get rid of the local use of `loading` and `error` states

sultanofcardio replied:
I started out with `useAsync`, but I would still end up having to call the `execute` function inside a `useEffect`. I moved away from it because this approach is simpler

alex-i-e commented on src/modules/Auth/pages/RecoverPassword/RecoverPassword.tsx:
```diff
@@ -0,0 +1,62 @@
+  if (error) {
+    return (
+      <Box display="flex" flex={1} justifyContent="center" alignItems="center" textAlign="center">
+        <StyledBodyLarge variant="body1" margin="16px 0px">
```
please consider to use `theme.spacing`

sultanofcardio replied:
Thanks for the suggestion. I ended up just removing the margin because it wasn't necessary. I copied this component from the respondent web app repo and modified it

alex-i-e commented on src/modules/Auth/pages/RecoverPassword/RecoverPassword.tsx:
```diff
@@ -0,0 +1,62 @@
+
+  if (error) {
+    return (
+      <Box display="flex" flex={1} justifyContent="center" alignItems="center" textAlign="center">
```
Maybe It's time to discuss the styling of components, but so far we stick the usage of `Styled[ComponentName]` as a common approach like you applied in [RecoverForm.styles.ts](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1510/files#diff-285908779dbf07e580c21b701a9afedcceb52cdb7c6d1c809e5c54cdb31cd4ca)
@IhorNesterenko99, please add your thoughts.

sultanofcardio replied:
I opted to use inline styles here because the `RecoverPassword` component is a page, and so far I've only seen the styled component approach used for components in the `features` directory. However, I agree that these styles could be abstracted. Do you have an opinion on where that file could go?

alex-i-e replied:
Yeap, make sense in this particular case.

sultanofcardio committed:
Fix password confirm error message

The password confirmation field should say 'Please confirm your password' when the field is empty, and only say 'Your passwords do not match' when the user has actually entered something

Co-authored-by: Aleksandr Eremin <aleksandr.i.eremin@yandex.ru>

sultanofcardio committed:
M2-4835: Remove inline margin and unnnecessary div
